doc/reference/security.rst

```diff
@@ -257,6 +257,64 @@ While formatting the template differently would prevent such vulnerabilities.
         font-weight: bold;
     }
 
+Escaping vs Sanitizing
+----------------------
+
+**Escaping** converts `TEXT -> CODE`. It is absolutely mandatory to do it every
+time you mix `DATA/TEXT` with `CODE` (e.g. generating HTML or python code to
+be evaluated inside a `safe_eval`), because `CODE` always requires `TEXT` to be
+encoded. It is critical for security, but it's also a question of correctness.
+Even when there is no security risk (because the text is 100% guarantee to be
+safe or trusted), it is still required (e.g. to avoid breaking the layout in
+generated HTML).
+
+Escaping will never break any feature, as long as the developer identifies which
+variable contains `TEXT` and which contains `CODE`.
+
+.. code-block:: python
+
+    >>> from odoo.tools import html_escape, html_sanitize
+    >>> data = "<R&D>" # `data` is some TEXT coming from somewhere
+
+    # Escaping turns it into CODE, good!
+    >>> code = html_escape(data)
+    >>> code
+    '&lt;R&amp;D&gt;'
+
+    # Now you can mix it with other code...
+    >>> self.message_post(body="<strong>%s</strong>" % code)
+
+**Sanitizing** converts `CODE -> SAFER_CODE`. It does not work on `TEXT`.
```

Comment:
Might be worth emphasizing that SAFER_CODE does not necessarily mean *safe* code, sanitization / validation can have holes and edge cases, which usually is not the case for escaping (this is kinda-sorta on the edge of [parse don't validate](https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/)). So in the spectrum between safe and unsafe, sanitisation is way closer to "unsafe" than escaping.

Reply:
(and sanitisation safety is balanced against sanitisation breakages: the safer the sanitisation the more likely it is to break things)